### PR TITLE
ESLint Plugin: Update eslint jsdoc dependency

### LINKED ISSUE #34297: Scripts: `wp-scripts lint-js` Breaks After v16.1.4
### Description

After updating `@wordpress/scripts` from `16.1.4` to `17.0.0` or higher (currently at `18.0.0`), the JS linter breaks with the following result:

```
$ wp-scripts lint-js

Oops! Something went wrong! :(

ESLint: 7.32.0

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/parser/tokenizers/description.js' is not defined by "exports" in /Users/dom/Projects/wp-scripts-lint-js-test-repo/node_modules/@es-joy/jsdoccomment/node_modules/comment-parser/package.json
    at throwExportsNotFound (internal/modules/esm/resolve.js:290:9)
    at packageExportsResolve (internal/modules/esm/resolve.js:513:3)
    at resolveExports (internal/modules/cjs/loader.js:437:36)
    at Function.Module._findPath (internal/modules/cjs/loader.js:477:31)
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:872:27)
    at Function.Module._load (internal/modules/cjs/loader.js:730:27)
    at Module.require (internal/modules/cjs/loader.js:957:19)
    at require (/Users/dom/Projects/wp-scripts-lint-js-test-repo/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at Object.<anonymous> (/Users/dom/Projects/wp-scripts-lint-js-test-repo/node_modules/@es-joy/jsdoccomment/dist/index.cjs.cjs:7:28)
    at Module._compile (/Users/dom/Projects/wp-scripts-lint-js-test-repo/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Upon initial investigation, I think the `@wordpress/eslint-plugin/jsdoc` preset may be causing this issue. It’s a part of `@wordpress/eslint-plugin/recommended` used by default in `@wordpress/scripts`’ ESLint config.

### Step-by-step reproduction instructions

Clone the [test repository](https://github.com/Automattic/wp-scripts-lint-js-test-repo), then run:

```sh
yarn
yarn format # works fine
yarn lint:js # will break
```

### Screenshots, screen recording, code snippet

_No response_

### Environment info

```
$ node -v
v14.17.0

$ cat package.json | grep @wordpress/scripts
		"@wordpress/scripts": "18.0.0",
```

Gutenberg not installed.

The test repository also includes `eslint-plugin-import`, but that’s primarily to keep the project config as identical to the project config as possible.

### Pre-checks

- [X] I have searched the existing issues.
- [X] I have tested with all plugins deactivated except Gutenberg.

### PR DESCRIPTION
**Pull request by renatho**
Fixes https://github.com/WordPress/gutenberg/issues/34297

## Description

* It updates the `eslint-plugin-jsdoc` in order to solve the reported issue.
* It also disables some rules that started failing after the update that wouldn't be easy to fix.

## How has this been tested?

`npm run lint-js` in Gutenberg repo.

## Types of changes

Breaking change

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/eslint-plugin/configs/jsdoc.js [hunk 1]
```diff
@@ -105,6 +105,9 @@ module.exports = {
 		'jsdoc/require-param-description': 'off',
 		'jsdoc/require-returns': 'off',
 		'jsdoc/require-yields': 'off',
+		'jsdoc/tag-lines': 'off',
+		'jsdoc/no-multi-asterisks': 'off',
+		'jsdoc/valid-types': 'off',
```
Comment (renatho) [inline packages/eslint-plugin/configs/jsdoc.js]: This rule was previously enabled, but there's a lot of cases where we're using invalid types. I didn't dig into that too much, but probably the rule was improved and it getting new cases. So, for now, I decided to just disable it.

Comment (gziolo) [reply packages/eslint-plugin/configs/jsdoc.js]: It looks like our setup aligns with the following rule configurations:

```js
/**
 * Some description
 * @param {string} a
 *
 * @param {number} b
 *
 */
// "jsdoc/tag-lines": ["error"|"warn", "never"]
// Message: Expected no lines between tags
```

```js
/**
 * Desc.
 *
 **/
// "jsdoc/no-multi-asterisks": ["error"|"warn", {"preventAtEnd":true}]
// Message: Should be no multiple asterisks on end lines.
```

I see that we already disable `jsdoc/valid-types` in a few places. I would be fine with disabling the rule for now and explore if we can find a good way to enable it without exceptions. Although maybe we don't need it given that TypeScript does it's own type validation in packages that have it enabled  🤷🏻‍♂️

**Comment by gziolo** [comment]
Nice one. Thank you for taking care of the reported issue so quickly. 

Should we consider it a breaking change for the ESLint plugin package (it would also cascade to `@wordpress/scripts` which use the recommended ESLint settings)? Are new rules set on the error level or they are just warnings?

File: packages/eslint-plugin/configs/jsdoc.js (hunk 1 above)
Comment (renatho) [reply packages/eslint-plugin/configs/jsdoc.js]: I updated the `no-multi-asterisks` rule (https://github.com/WordPress/gutenberg/pull/34338/commits/0201bbff5b94e409141a50104768a8514667fda9). But for the `tag-lines` I tried some different combinations with the `tags` and `noEndLines` options, but it seems we can't reproduce the scenario we use:

```js
/**
 * Some description.
 * 
 * @param {string} a
 * @param {number} b
 * 
 * @return {number} c
 */
```

**Comment by renatho** [comment]
> Nice one. Thank you for taking care of the reported issue so quickly.
> 
> Should we consider it a breaking change for the ESLint plugin package (it would also cascade to `@wordpress/scripts` which use the recommended ESLint settings)? Are new rules set on the error level or they are just warnings?

Thank you for the quick review! You're right, it should probably be considered as a breaking change (updated in the PR description).
For Gutenberg repo, with the update rules, it's with the same warnings, but it can have different impact in other projects (and we also changed some rules, like disabling the `jsdoc/valid-types`).

File: packages/eslint-plugin/configs/jsdoc.js [hunk 2]
```diff
@@ -133,6 +139,5 @@ module.exports = {
 		'jsdoc/require-returns-check': 'error',
 		'jsdoc/require-returns-description': 'error',
 		'jsdoc/require-returns-type': 'error',
-		'jsdoc/valid-types': 'error',
```
Comment (gziolo) [inline packages/eslint-plugin/configs/jsdoc.js]: So it looks like we use this rule for all non-TypeScript files:

https://github.com/WordPress/gutenberg/blob/7901895ca20cf61e402925e31571d659dab64721/.eslintrc.js#L241-L245

How many new errors does it report if we leave this rule enabled? Maybe we should fix them?

Everything else look good.

File: packages/eslint-plugin/package.json [hunk 3]
```diff
@@ -39,7 +39,7 @@
 		"eslint-config-prettier": "^7.1.0",
 		"eslint-plugin-import": "^2.23.4",
 		"eslint-plugin-jest": "^24.1.3",
-		"eslint-plugin-jsdoc": "^34.1.0",
+		"eslint-plugin-jsdoc": "^36.0.8",
```
Comment (gziolo) [inline packages/eslint-plugin/package.json:42]: As far as I can tell the only braking change that impacts the WordPress plugin is from v35:

https://github.com/gajus/eslint-plugin-jsdoc/compare/v34.8.2...v35.0.0

> `multiline-blocks`: Should have been enabled by default

See https://github.com/gajus/eslint-plugin-jsdoc/commit/b81e25d14d05c5c65992c9e6f34386fdf14e8d8d.

If we set `jsdoc/multiline-blocks` and `jsdoc/no-multi-asterisks` at the warning level, the we can consider this PR as adding new features only. What do you think?

File: packages/eslint-plugin/configs/jsdoc.js (hunk 2 above)
Comment (renatho) [reply packages/eslint-plugin/configs/jsdoc.js]: It wasn't too much. I fixed some, and I just disabled others here: https://github.com/WordPress/gutenberg/pull/34338/commits/ffec4db403daf875d18ca89a56fef38673b1f138

The ones I disabled were using `{0|-1|null}` as type. As far as I found, we can have `{(1|2)}` (not negative) and `{('a'|'b')}`, but I didn't find how to combine a mix.

File: packages/eslint-plugin/package.json (hunk 3 above)
Comment (renatho) [reply packages/eslint-plugin/package.json:42]: I think you meant https://github.com/gajus/eslint-plugin-jsdoc/compare/v34.8.2...v36.0.0, right? Or do you think we should update until the v35 only?

Comment (gziolo) [reply packages/eslint-plugin/package.json:42]: v36 dropped only support for Node 10 or something like that which isn't relevant in our context. It's fine as is 👍🏻

File: packages/eslint-plugin/configs/jsdoc.js (hunk 2 above)
Comment (gziolo) [reply packages/eslint-plugin/configs/jsdoc.js]: Knowing how that works, it probably could work as a type defined separately and used in those places. Anyway, it eventually become non-issue when we migrate to TypeScript.

**Review by gziolo — APPROVED** [review]
I included a commit that adds an entry to the CHANGELOG file for `@wordpress/eslint-plugin` package and this PR should be good to go when CI turns green. Nice work, thank you for going through all the feedback shared 🙇🏻

**Comment by gziolo** [comment]
It looks like this branch will need a rebase before we can proceed.

**Comment by renatho** [comment]
> It looks like this branch will need a rebase before we can proceed.

Thank you for the review @gziolo! =)
Branch rebased and fixed a new warning after the rebase: https://github.com/WordPress/gutenberg/pull/34338/commits/d941a6d4f879a6cfd235b546a82502b646eaab8c

**Comment by gziolo** [comment]
🎉 Let's merge when CI turns green 👍🏻

**Comment by gziolo** [comment]
I see CI failures for the RN checks so we can ignore them here.

**Pull request merged**